### PR TITLE
WS-140 - Add getLastNFTId

### PR DESCRIPTION
**Pull request by AgusVelez5**
**Type of MR:**

- [x] Feature
- [ ] Fix
- [ ] HotFix

**Small description:**

- Add getLastNFTId
- Add ManagerMustActive decorator
- Refactor NFTHandler

**Aditional observations/considerations:**
...

# Checklist:

- [x] My code follows the styleguides of the project (successfully pass linters)
- [x] I tested my code.
- [x] I commented/docummented my code in a way that complex parts are explained.
- [ ] If required, i updated project's README file.
- [x] I added/updated tests that tests my changes in code.
- [x] All previous and new tests pass.
- [ ] I commit any new dependency.
- [ ] I updated UML diagram in confluence if any model was modified in this MR.

### CODE REVIEW

File: packages/nft/src/NFTHandler.ts [hunk 1]
```diff
@@ -73,61 +77,51 @@ export class NFTHandler {
...
 
     return await Promise.all(
       tokenIds.map(tokenId => this.get(nftManager, storage, tokenId, nftFormat))
     )
   }
 
+  /**
+   * It takes a owner, and returns a list of NFT objects
+   * @param {NFT721Manager} nftManager - NFT721Manager - The NFT721Manager instance that will be used
+   * to fetch the NFT tokenURI.
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:102]: to fetch the NFT tokenURI? it should be the tokenId

File: packages/nft/src/NFTHandler.ts [hunk 2]
```diff
@@ -240,4 +235,70 @@ export class NFTHandler {
...
+      })
+      return value
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
+      throw BALANCE_OF_FAILED(address, error)
+    }
+  }
+
+  /**
+   * Takes an address and an index and returns the corresponding tokenUri.
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:272]: the corresponding token ID

File: packages/nft/src/NFTHandler.ts [hunk 3]
```diff
@@ -240,4 +235,70 @@ export class NFTHandler {
...
+   * @param {NFT721Manager} nftManager - NFT721Manager - The NFT721Manager instance that will be used
+   * to fetch the tokenUri.
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:274]: to fetch the token ID

File: packages/nft/src/NFTHandler.ts [hunk 4]
```diff
@@ -240,4 +235,70 @@ export class NFTHandler {
...
+   * @param {string} owner - The owner of the nft.
+   * @param {number} index - The index of the tokenUri.
+   * @returns The balance
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:277]: returns the token ID for the provided index and owner

File: packages/nft/src/NFTHandler.ts [hunk 5]
```diff
@@ -240,4 +235,70 @@ export class NFTHandler {
...
+
+    try {
+      const { value } = await nftManager.execute({
+        method: 'tokenOfOwnerByIndex(address,uint256)',
+        params: [owner, index]
+      })
+      return value
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
+      throw TRANSACTION_FAILED(error)
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:301]: I think that the execute already throws an error if the execute fails, is it necessary to try/catch it?

File: packages/nft/src/NFTHandler.ts [hunk 6]
```diff
@@ -11,12 +11,15 @@ import { NFTJsonFactory } from './NFTJsonFactory'
 import { NFTImageFactory } from './NFTImageFactory'
 import { NFTJsonImageFactory } from './NFTJsonImageFactory'
 
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { ManagerMustActive } from './utils'
```
[the remark this answers is not in the document]
Comment (alfarkas) [reply packages/nft/src/NFTHandler.ts:15]: I like the second one.

File: packages/nft/src/NFTHandler.ts [hunk 7]
```diff
@@ -73,61 +77,51 @@ export class NFTHandler {
...
     owner: string,
     nftFormat: NFT_METADATA_FORMAT
   ): Promise<NFT[]> {
-    if (!nftManager.implements('tokenOfOwnerByIndex', ['address', 'uint256'])) {
-      throw FUNCTION_NOT_IMPLEMENTED(
-        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-        nftManager.contractAddr!,
-        'tokenOfOwnerByIndex(address,uint256)'
+    const ownerBalance = await NFTHandler.getAddressBalance(nftManager, owner)
+    const tokenIdexs = [...Array(ownerBalance).keys()]
```
Comment (alfarkas) [inline packages/nft/src/NFTHandler.ts:86]: typo? "tokenIndexes"? "tokenIds"?

File: packages/nft/src/NFTHandler.ts [hunk 8]
```diff
@@ -240,4 +235,70 @@ export class NFTHandler {
     }
     return tokenUri
   }
+
+  /**
+   * It takes an address and returns his balance
```
Comment (alfarkas) [inline packages/nft/src/NFTHandler.ts:239]: "its balance" instead of "his balance"

File: packages/nft/src/NFTHandler.ts [hunk 9]
```diff
@@ -32,7 +31,7 @@ export class NFTHandler {
    * @param {NFT_METADATA_FORMAT} nftFormat - The format of the NFT metadata.
    * @returns The NFT object
    */
-  @ManagerMustActive
+  @ManagerMustBeActive()
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:34]: missing the parameter => @ManagerMustBeActive('nftManager')

File: packages/nft/src/NFTHandler.ts [hunk 10]
```diff
@@ -243,7 +242,7 @@ export class NFTHandler {
    * @param {string} address - The address of which we want the balance.
    * @returns The balance
    */
-  @ManagerMustActive
+  @ManagerMustBeActive()
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:245]: same here (missing parameter)

File: packages/nft/src/NFTHandler.ts [hunk 11]
```diff
@@ -269,14 +268,14 @@ export class NFTHandler {
...
    * @param {NFT721Manager} nftManager - NFT721Manager - The NFT721Manager instance that will be used
-   * to fetch the tokenUri.
+   * to fetch the tokenId.
    * @param {string} owner - The owner of the nft.
    * @param {number} index - The index of the tokenUri.
-   * @returns The balance
+   * @returns The token ID for the provided index and owner
    */
-  @ManagerMustActive
+  @ManagerMustBeActive()
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:278]: same here (missing parameter)

File: packages/nft/src/NFTHandler.ts (hunk 7 above)
Comment (AgusVelez5) [reply packages/nft/src/NFTHandler.ts:86]: yep, typo

File: packages/nft/src/NFTHandler.ts [hunk 12]
```diff
@@ -97,11 +96,11 @@ export class NFTHandler {
   }
 
   /**
-   * It takes a owner, and returns a list of NFT objects
+   * It takes a owner, and returns the last tokenId.
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:99]: also this should be "It takes an owner"

File: packages/nft/src/NFTHandler.ts [hunk 13]
```diff
@@ -240,4 +234,65 @@ export class NFTHandler {
...
+      const { value } = await nftManager.execute({
+        method: 'balanceOf(address)',
+        params: [address]
+      })
+      return value
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    } catch (error: any) {
+      throw BALANCE_OF_FAILED(address, error)
+    }
+  }
```
Comment (mauricio-b-r) [inline packages/nft/src/NFTHandler.ts:268]: is the try/catch needed?

Comment (AgusVelez5) [reply packages/nft/src/NFTHandler.ts:268]: It depends, to give a little clearer error to the dev.

**Review by alfarkas — APPROVED** [review]

**Review by mauricio-b-r — APPROVED** [review]

**Pull request merged**